logic/path/src/channel_graph.rs

```diff
@@ -189,58 +264,104 @@ impl ChannelGraph {
             );
             Some(ret)
         } else {
-            let weighted = ChannelEdge { channel, quality: None };
-            self.graph.add_edge(channel.source, channel.destination, weighted);
+            // Otherwise, create a new edge and add the nodes with 0 quality if they don't yet exist
+            let src = *self.indices.entry(channel.source).or_insert_with(|| {
+                self.graph
+                    .add_node(Node {
+                        address: channel.source,
+                        quality: 0.0,
+                    })
+                    .index() as u32
+            });
+
+            let dst = *self.indices.entry(channel.destination).or_insert_with(|| {
+                self.graph
+                    .add_node(Node {
+                        address: channel.destination,
+                        quality: 0.0,
+                    })
+                    .index() as u32
+            });
+
+            let weighted = ChannelEdge { channel, score: None };
+
+            self.graph.add_edge(src.into(), dst.into(), weighted);
             debug!("new {channel}");
 
             None
         }
     }
 
-    /// Updates the quality value of network connection between `source` and `destination`
-    /// The given quality value must be always non-negative
-    pub fn update_channel_quality(&mut self, source: Address, destination: Address, quality: f64) {
+    /// Updates the quality of a node (inserting it into the graph if it does not exist yet).
+    /// The given quality value must always be non-negative.
+    pub fn update_node_quality(&mut self, address: &Address, quality: f64) {
         assert!(quality >= 0_f64, "quality must be non-negative");
-        if let Some(channel) = self.graph.edge_weight_mut(source, destination) {
-            if quality != channel.quality.unwrap_or(-1_f64) {
-                channel.quality = Some(quality);
-                debug!("updated quality of {} to {quality}", channel.channel);
+        if !self.me.eq(address) {
+            match self.indices.entry(*address) {
+                // The node exists and we're updating to greater-than-zero quality
+                Entry::Occupied(existing) => {
+                    let existing_idx: NodeIndex = (*existing.get()).into();
+                    if quality > 0.0 || self.graph.neighbors_undirected(existing_idx).count() > 0 {
```

Comment:
Why not remove the node when the quality is below the configured threshold already?